7_react/react1/week3/vite-project/src/App.jsx

```diff
@@ -0,0 +1,85 @@
+import React, { useEffect, useState } from "react";
+import "./App.css";
+
+import AddTodo from "./components/AddTodo";
+import TodoList from "./components/TodoList";
+import Timer from "./components/Timer";
+
+const App = () => {
+  const [todos, setTodos] = useState([]);
+  useEffect(() => {
+    fetch(
+      "https://gist.githubusercontent.com/benna100/391eee7a119b50bd2c5960ab51622532/raw"
+    )
+      .then((response) => response.json())
+      .then((data) => {
+        setTodos(data);
+        data.map((todo) => ({
+          id: todo.id,
+          description: todo.description,
+          deadline: todo.deadline,
+          done: todo.done,
+        }));
+      });
+  }, []);
+
+  const addTodo = (newTodo) => {
+    if (!newTodo.description || !newTodo.deadline) {
+      alert("Please add description and deadline");
+      return;
+    }
+    const deadlineDate = new Date(newTodo.deadline);
+    const today = new Date();
+    if (deadlineDate < today) {
+      alert("Deadline cannot be in the past");
+      return;
+    }
+    setTodos((prevTodos) => [
+      ...prevTodos,
+      { id: prevTodos.length + 1, ...newTodo, deadline: newTodo.deadline },
+    ]);
+  };
+
+  const toggleDone = (id) => {
+    setTodos(
+      todos.map((todo) =>
+        todo.id === id ? { ...todo, done: !todo.done } : todo
+      )
+    );
+  };
+
+  const deleteTodo = (id) => {
+    setTodos(todos.filter((todo) => todo.id !== id));
+  };
+
+  const editTodo = (id, newDescription, newDeadline) => {
+    setTodos(
+      todos.map((todo) =>
+        todo.id === id
+          ? { ...todo, description: newDescription, deadline: newDeadline }
+          : todo
+      )
+    );
+  };
+
+  return (
+    <div className="app">
+      <Timer />
+      <h1 className="app_title">What I need ToDo? &#129300;</h1>
+      <p className="app_description">Create your own list</p>
+
+      {todos.length === 0 ? (
+        <p>No items</p>
+      ) : (
+        <TodoList
+          todos={todos}
```

Comment:
The prop types for `<TodoList>` requires each todo to have a boolean containing whether it is done. This is not the case for the todos provided by the `fetch`. Perhaps some processing should be done to those fetched todos or the prop types should be modified. What do you think is a good solution? 😄 

This also causes some issues with your checkboxes as React assumes them to be uncontrolled when their value is undefined, but then they get a value when you check them off. [You can read more about this issue here.](https://react.dev/reference/react-dom/components/input#im-getting-an-error-a-component-is-changing-an-uncontrolled-input-to-be-controlled)

Reply:
Nice with some additional error handling and an easier to read async function! I think you are on the right track with the `loading` state, but you are not using the value anywhere at the moment. If you only show the todo list once data has loaded, then we get around the issue. 😄